Main settings page persistence

waclawjacek opened the pull request:
# Changes proposed in this Pull Request:

Fixes #1812.

This PR adds data API handling of all the settings on the WooCommerce > Settings > Payments > Stripe UPE page.

Clicking checkboxes or inputting values will update the `@wordpress/data` store, and clicking "Save changes" will submit a request to the back-end REST controller.

## A change in approach to settings store actions, hooks, and selectors

Since all of the selectors were very similar in function, we decided to simplify things by removing selectors and moving their logic into hooks. I also stole the idea to replace the actions with direct calls to `updateSettingsValues()` (introduced by @frosso in another PR). Since neither selectors or actions were no longer needed, all the logic (and future scope of changes) for retrieving and updating each setting is now limited to just `client/data/settings/hooks.js`.

I have also replaced the separate definitions for each hook with calls to a creator function - `makeSettingsHook()`.

I have also simplified tests by running the same set of tests for each hook.

Hopefully this will make it easier to add new hooks and discover bugs.

## REST controller test simplification

I have also simplified the REST controller tests (and added tests for untested fields) by extracting the shared test logic (similarly to the JS changes above) by field type (boolean, enum, or statement descriptor which has its own validator callback).

This allows for quickly adding new tests based on field type and making sure all the appropriate tests are run.

# Testing instructions

1. Enable UPE preview and UPE.
1. Go to the UPE settings page.
1. A loading placeholder should be shown while settings are loaded.
1. Change settings and save them.
1. Refresh the page - settings should be persisted and a snackbar notification confirming success should appear.
1. Disable UPE and confirm that the settings are presented with the same values in the old UI.

Commit:
Add support for saving payment request button locations

Commit:
Add support for saving General card on Settings page

Commit:
Add support for saving Payments & transactions card on Settings page

Commit:
Add support for saving Debug mode on Settings page

Commit:
Replace repetitive hook definitions with hook creator calls

Props frosso.

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

harriswong commented:
Merged `develop` and attempted to resolve conflicts introduced by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1847. 

After merged, the express section stopped saving. Looking into it.

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Rename payment_request_enabled_locations to payment_request_button_locations

Commit:
Remove duplicated call to set express checkout enabled

Commit:
Remove unused wcpay related codes

Commit:
npm run lint:php-fix

Commit:
Changing unit test to integration test to match the rest of the file

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Update PaymentsAndTransactionsSection tests to use wcstripe/data

Commit:
Use wcstripe/data in <DebugMode>

Commit:
Remove outdated tests

Commit:
Replace usePaymentRequest* with generated functions. Extract test logic

Replace usePaymentRequest*() hooks with ones generated using
makeReadWritePairHook().

Extract common hook test logic to a creator function.

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Generate useEnabledPaymentMethodIds using creator

Commit:
Wrap action in hook creator in useCallback()

Commit:
Memoize mapSelect in generated hooks

Commit:
Refactor generated hook tests using data provider pattern

Commit:
Add tests for all generated hooks

Commit:
Add updateSettingsValues() tests

Commit:
Revert useDevMode() to read-only hook

Commit:
Add dev mode support to GeneralSettingsSection

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Move "is enabled" methods to WC_Gateway_Stripe and update typehint

Commit:
Add missing REST update_settings args

Commit:
Port WCPay REST controller tests. Extract UPE_Utils trait for tests

Commit:
Reorder and add missing update_settings args specs

Commit:
Extract shared bool and enum test logic. Fix discovered bugs

Implementing the shared logic revealed a bug in handling
is_debug_log_enabled and payment request settings.

Commit:
Extract statement descriptor test logic

Commit:
Merge branch 'develop' into add/1812-settings-persistence-main-settings-page

Commit:
Rename hook generator functions. Introduce makeReadOnlySettingsHook()

Commit:
Update WC_Stripe_Test to use `self` to reference UPE_Utils

Commit:
Port GeneralSettingsSection tests from WCPay

Commit:
Fix makeReadOnlySettingsHook()

Commit:
Fix validation for short statement descriptor

waclawjacek commented on client/data/settings/__tests__/hooks.js:
```diff
@@ -155,102 +145,148 @@ describe( 'Settings hooks tests', () => {
+		},
+	};
+
+	describe.each( Object.entries( generatedHookExpectations ) )(
```
This is a set of tests that will be ran for each hook from the above variable. Since all of the settings should behave the same way, it made sense to group them here.

waclawjacek commented on client/data/settings/hooks.js:
```diff
@@ -4,6 +4,56 @@ import { STORE_NAME } from '../constants';
 
 const EMPTY_ARR = [];
 
+const makeSettingsHookFromUpdateHandler = (
```
Creates a `[ value, updateCallback ]` pair that almost all of the hooks return.

waclawjacek commented on client/data/settings/hooks.js:
```diff
@@ -4,6 +4,56 @@ import { STORE_NAME } from '../constants';
+	return [ field, handler ];
+};
+
+const makeSettingsHook = ( fieldName, fieldDefaultValue = false ) => {
```
Same as above but with a default update callback that just overwrites the setting value with a new one. This approach is used by most of the hooks.

waclawjacek commented on client/data/settings/hooks.js:
```diff
@@ -4,6 +4,56 @@ import { STORE_NAME } from '../constants';
+	);
+};
+
+const makeReadOnlySettingsHook = (
```
Same as above but only returns a value for reading (no update callback).

waclawjacek commented on client/data/settings/hooks.js:
```diff
@@ -31,17 +81,59 @@ export const useSettings = () => {
-
-	return [ isPaymentRequestEnabled, updateIsPaymentRequestEnabled ];
-};
+export const useEnabledPaymentMethodIds = makeSettingsHook(
```
I'm fairly certain these cannot be created by iterating a list of names because export names have to be explicit.

waclawjacek commented on client/settings/payment-settings/general-settings-section.js:
```diff
@@ -31,12 +33,21 @@ const GeneralSettingsSection = () => {
+					disabled={ isDevModeEnabled }
+					onChange={ setIsTestModeEnabled }
+					label={
+						isDevModeEnabled
```
Dev mode is currently not supported by the plugin AFAIK.

frosso replied:
I don't think it is either - I created this to identify loose ends: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1727

Should we remove the `useDevMode()` hook and all its usage?

waclawjacek replied:
I think so! Best not keep it because if someone starts using it in the future and writes some code to handle it, it will also do things X and thing Y which are already implemented.

Removed in 4248c6cf8079c2938f335a08e6e084b0adb09e48.

frosso replied:
That's also a good point that I hadn't considered 👍

waclawjacek commented on tests/phpunit/test-wc-stripe.php:
```diff
@@ -96,37 +98,12 @@ public function statement_descriptor_sanitation_provider() {
 		];
 	}
 
-	private function enableUpeFeatureFlag() {
```
Extracted to the `UPE_Utils` trait.

frosso commented on client/data/settings/__tests__/hooks.js:
```diff
@@ -155,102 +145,148 @@ describe( 'Settings hooks tests', () => {
-				usePaymentRequestButtonTheme()
-			);
-			const [ value ] = result.current;
+	const generatedHookExpectations = {
```
(minor - no change required) This is very clever, but I'm afraid it might get tricky to maintain in the long run.
In my personal experience, when tests are "generated" this way, it gets harder to debug when something breaks. It takes time to parse what the code is doing and also what the tests are doing, sometimes even jumping up and down on the test file.
Being DRY can be a decent practice in a codebase, but when it comes to tests a WET practice could be more beneficial to parse.
Not saying that we can't write utilities to be reused across tests, just that when there are too many utilities in one test module, things can become harder to parse. At which point do we start writing the tests for the generated tests? 😄 

Since all the hooks are using the same "generator", testing the following might be sufficient:
- Test that the handler updates the attribute on one hook
- Test the "selected" value of one hook
- Test the fallback value for the selector of one hook with an array as default
- Test the fallback value for the selector of one hook with a string as default

That way we'd be testing that the hook grabs the correct attribute and that it returns the correct fallback value.

LMK your thoughts - have you ever had trouble with "generated" tests?

waclawjacek replied:
Thanks! I agree about the point about tests being too complicated and I think this applies to this PR.

As a for the reason why I decided to go with a generated approach, I felt overwhelmed by the number of tests that we had, and they all basically tested the same things (or could do that). I wanted to make sure all of the hooks are tested but there was no way to guarantee that other than going through the list of test functions and checking if there's a test for, say, an update action executing correctly. Even if there was one, there could be some tiny bugs in the tests, or if something changes in the tested code, you'd have to apply the same update to each test.

I think part of the problem here is that the way I grouped the tests is a bit too convoluted. E.g.:

```js
describe.each( Object.entries( generatedHookExpectations ) )(
		'%s()',
```

🙁 

> At which point do we start writing the tests for the generated tests? 😄

:smile: 100%.

> Since all the hooks are using the same "generator", testing the following might be sufficient:

My thinking here was that a specific hook can be, but doesn't have to be, created using the `makeSettingsHook()` function. Avoiding exposing it to the outside world and just testing the hooks assuming nothing guarantees that the hook behaves as expected no matter the underlying code.

> LMK your thoughts - have you ever had trouble with "generated" tests?

I can't recall any situations where I had to struggle with generated tests but just the fact that you have to spend time trying to understand what's going on in test code is not a good sign.

I am leaving this as it is for now because I'd like to get your opinion on the above and not because I believe this is the absolute right call. :smile:

frosso replied:
Thanks for going over the explanation with me! After re-reading my comment I feel like I might have sounded harsh, especially with the "At which point do we start writing the tests for the generated tests?" statement. My apologies for that! 😊 

> As a for the reason why I decided to go with a generated approach, I felt overwhelmed by the number of tests that we had, and they all basically tested the same things (or could do that).

I can definitely see that - we do have a lot of hooks!

> My thinking here was that a specific hook can be, but doesn't have to be, created using the makeSettingsHook() function. Avoiding exposing it to the outside world and just testing the hooks assuming nothing guarantees that the hook behaves as expected no matter the underlying code.
> [...]
> I am leaving this as it is for now because I'd like to get your opinion on the above and not because I believe this is the absolute right call. 😄

This sounds good, and you definitely put a lot of work on ensuring that the tests cover all of our bases, which I appreciate!
To be honest, I am _not_ one of the people that promotes 100% test coverage for all of our tests, so I might be a bit biased here :) it is indeed accurate that a hook can (or not) be created with the `makeSettingsHook()` function. However most of the current ones are (with the exclusion of `useSettings`), so (still in my opinion) it would be more than acceptable to just test a few scenarios for one (or two) of the generated hooks (without exposing `makeSettingsHook()`). If new hooks were to be added, and those hooks didn't use `makeSettingsHook()`, we could just ask to write the tests for them separately :)
After all, the code to be covered in tests is in `makeSettingsHook()`. Assigning a constant like `export const newHook = makeSettingsHook('new_attribute', defaultValue)` is not _really_ code coverage - the code is executed in `makeSettingsHook()` at the import, so I don't believe it would be highlighted by a static analysis tool. Only the body of `makeSettingsHook()` would be, if I understand it correctly.

Again, no change necessary here, I appreciate the effort you put into making sure that all of our bases are covered! 💪

waclawjacek replied:
> After re-reading my comment I feel like I might have sounded harsh, especially with the "At which point do we start writing the tests for the generated tests?" statement. My apologies for that! 😊

No worries! I did not read it as such but thank you so much for saying this!

> However most of the current ones are (with the exclusion of useSettings), so (still in my opinion) it would be more than acceptable to just test a few scenarios for one (or two) of the generated hooks (without exposing makeSettingsHook()). If new hooks were to be added, and those hooks didn't use makeSettingsHook(), we could just ask to write the tests for them separately :)

I usually try to test as much as possible but I think I went overboard with hypothetical issues here rather than, as you said, focusing on just testing all the possible flows in `makeSettingsHook()` once.

Thanks so much for this discussion! I am leaning towards leaving this as-is just for the sake of getting this PR merged but I am definitely 100% fine with changing the tests to something simpler. WDYT?

---

I meant to write this in my previous comment but I forgot! Going with the test *all the things* approach allowed me to [find some bugs in the existing code](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1852/commits/40f7c58ebcbd0dfb0349ca265ead5e6fc36f8c60). This was in PHP code though, not JS, but I took a similar approach there where I defined a list of "field specs" to validate fields against.

frosso replied:
> I am leaning towards leaving this as-is just for the sake of getting this PR merged but I am definitely 100% fine with changing the tests to something simpler. WDYT?

Totally good, let's merge what you have here 👍

frosso commented on client/data/settings/__tests__/reducer.js:
```diff
@@ -72,218 +68,115 @@ describe( 'Settings reducer tests', () => {
+				quuz: 'corge',
+			} );
+		} );
+		test( 'overwrites existing settings in the `data` subtree', () => {
```
(minor) There's different convention on spacing in this file - should we add some rule to enforce a new line between `test`/`it` blocks?

waclawjacek replied:
Whoops, good catch! I have no idea how this happened. Maybe I used some autoformatting? Fixed in ac0c6fcd6266053a44799fb4822c111c0dd17494.

> should we add some rule to enforce a new line between test/it blocks?

That'd definitely be useful but do you think it's worth the hassle?

frosso replied:
No worries - I was looking into a rule for this, but I couldn't find one. Maybe with a mix of existing rules we could get closer - let's just keep monitoring it :)

frosso commented on client/settings/payment-settings/__tests__/general-settings-section.test.js:
```diff
@@ -1,26 +1,100 @@
+			if ( ! isEnabled ) {
+				expectation = expectation.not;
+			}
+			expectation.toBeChecked();
```
Also in this case - similar to what I noted in the tests above. At which point are we going to have to start creating tests for the tests we're creating? 😄

frosso commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -27,11 +27,9 @@ class WC_REST_Stripe_Settings_Controller extends WC_Stripe_REST_Base_Controller
+	 * @param WC_Gateway_Stripe $gateway Stripe payment gateway.
 	 */
-	public function __construct( WC_Stripe_Payment_Gateway $gateway ) {
+	public function __construct( WC_Gateway_Stripe $gateway ) {
```
👍

frosso commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -58,62 +56,151 @@ public function register_routes() {
 				'args'                => [
-					'is_payment_request_enabled'       => [
-						'description'       => __( 'If Stripe express checkouts should be enabled.', 'woocommerce-gateway-stripe' ),
+					'is_stripe_enabled'                     => [
```
It looks like this hasn't been catched by the linter? 👀

waclawjacek replied:
Strangely though, after fixing (in 10560256da63170a033860956a84a4805f1853cc) to what phpcs wants it to be, this is what it says to be OK despite the larger number of spaces than in other lines...

![Markup on 2021-10-07 at 21:15:36](https://user-images.githubusercontent.com/1759681/136448413-3621e81f-8eb6-4c90-9b9a-b6b708813b36.png)

frosso commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -122,14 +209,31 @@ public function register_routes() {
+				'payment_request_button_locations'      => $this->gateway->get_option( 'payment_request_button_locations' ),
+
+				/* Settings > Payments & transactions */
+				'is_manual_capture_enabled'             => 'no' === $this->gateway->get_option( 'capture' ),
```
(minor) There's a `$WC_Stripe_UPE_Payment_Gateway->is_automatic_capture_enabled()` method on the UPE gateway - maybe in the future we can reuse that, by putting it in the `WC_Gateway_Stripe` class 👍

waclawjacek replied:
I have moved this method to `WC_Gateway_Stripe` and the REST controller is now calling it. Done in 4af737d073bd7a762bcde827066270cc5da33e6d.

frosso commented on tests/phpunit/utils/trait-upe-utils.php:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Provides methods useful when testing UPE-related logic.
+ */
+trait UPE_Utils {
```
Good idea! Should this be `UPE_Test_Utils`, to ensure that it's only used in tests?
Or maybe now that I think about it, having it in the `tests` directory is sufficient 🤷

waclawjacek replied:
I suppose, but good idea! Less ambiguity when using IDE autocompletion. Done in 0f959782c2f1244e1561613bec852b1325a1d564!

frosso commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -58,62 +56,151 @@ public function register_routes() {
+						'type'              => 'boolean',
+						'validate_callback' => 'rest_validate_request_arg',
+					],
+					'short_statement_descriptor'            => [
+						'description'       => __( 'We\'ll use the short version in combination with the customer order number.', 'woocommerce-gateway-stripe' ),
+						'type'              => 'string',
+						'validate_callback' => [ $this, 'validate_short_statement_descriptor' ],
+					],
```
I'm getting a validation error on this.
When I _don't_ provide a "short" statement descriptor and the "Add customer order number to the bank statement" checkbox is un-checkeched, the validation fails.
Should it be failing, since there is no short statement descriptor?
![Screen Shot 2021-10-06 at 2 15 31 PM](https://user-images.githubusercontent.com/273592/136268340-184ded6c-f385-4f17-b425-79ac879eb57b.png)

frosso replied:
As we discussed during our meeting, since this PR touches a decent number of files I created a separate ticket for this scenario: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2047

This will allow this PR to get merged, once the conflicts with the base branch have been resolved.

waclawjacek replied:
Great catch! As discussed during meeting, will fix in #2047.

frosso commented on client/data/settings/hooks.js:
```diff
@@ -4,6 +4,56 @@ import { STORE_NAME } from '../constants';
+) => () => {
+	const { updateSettingsValues } = useDispatch( STORE_NAME );
+
+	const field = useSelect(
+		( select ) => {
+			const { getSettings } = select( STORE_NAME );
+
+			return getSettings()[ fieldName ] || fieldDefaultValue;
+		},
+		[ fieldName, fieldDefaultValue ]
+	);
```
LMK if I'm missing something, but could this be replaced by `makeReadOnlySettingsHook`?
```
const value = makeReadOnlySettingsHook( fieldName, fieldDefaultValue );
```

waclawjacek replied:
Oh wow, great catch! :100: Added in ffdac2e8249a99b070f4049f0103979745b5f357.

frosso commented on client/data/settings/hooks.js:
```diff
@@ -31,17 +81,59 @@ export const useSettings = () => {
+);
+export const useDebugLog = makeSettingsHook( 'is_debug_log_enabled' );
+
+export const usePaymentRequestLocations = makeSettingsHookFromUpdateHandler(
+	'payment_request_button_locations',
+	( v, updateSettingsValues ) =>
+		updateSettingsValues( {
+			payment_request_button_locations: [ ...v ],
+		} ),
+	EMPTY_ARR
+);
```
Why can't this use `makeSettingsHook` directly? The value is still being updated in the redux store. And the array reference is already changing in the `PaymentRequestSection` component.

If we used `makeSettingsHook`, we could get away with just having a `makeReadOnlySettingsHook` (for read only) and `makeSettingsHook`, without the need of a third "factory" function

waclawjacek replied:
You're right! :100: Done in d013effd2226247a214c806a5b751f60912a9820.

frosso reviewed (commented):
Great job! I only hit one snag when testing the "Shortened customer bank statement" field.
The rest of the fields seem to be working great!
Should we also remove the `useDevMode` hook?

Commit:
Remove mentions of dev mode which the plugin does not support

Commit:
Fix spacing in client/data/settings/__tests__/reducer.js

Commit:
Fix REST controller args array indentation

Commit:
Call is_automatic_capture_enabled() instead of accessing option manually

Commit:
Rename UPE_Utils to UPE_Test_Utils

Commit:
Use makeReadOnlySettings() hook in makeSettingsHookFromUpdateHandler()

Commit:
Remove redundant makeSettingsHookFromUpdateHandler() function

Commit:
Merge remote-tracking branch 'origin/develop' into add/1812-settings-persistence-main-settings-page

Commit:
Replace UPE trait with class for PHP 5.6 compatibility

frosso reviewed (approved):
Thanks for the discussion and the changes and seeing this through the end! I tested the settings, their persistence and ensuring they are reflected in the "old" settings.

Looks good! 🚢
